crates/typst/src/math/stretch.rs

```diff
@@ -1,12 +1,151 @@
 use ttf_parser::math::{GlyphAssembly, GlyphConstruction, GlyphPart};
 use ttf_parser::LazyArray16;
 
-use crate::layout::{Abs, Frame, Point, Size};
-use crate::math::{GlyphFragment, MathContext, Scaled, VariantFragment};
+use crate::diag::SourceResult;
+use crate::foundations::{elem, Content, Packed, Resolve, Smart, StyleChain};
+use crate::layout::{Abs, Axis, Frame, Length, Point, Rel, Size, VAlignment};
+use crate::math::{
+    GlyphFragment, LayoutMath, MathContext, MathFragment, Scaled, VariantFragment,
+};
+use crate::utils::Get;
 
 /// Maximum number of times extenders can be repeated.
 const MAX_REPEATS: usize = 1024;
 
+/// Stretches a glyph.
+///
+/// This function can also be used to automatically stretch the base of an
+/// attachment, so that it fits the top and bottom attachments.
+///
+/// Note that only some glyphs can be stretched, and which ones can depend on
+/// the math font being used. However, most math fonts are the same in this
+/// regard.
+///
+/// ```example
+/// $ H stretch(=)^"define" U + p V $
+/// $ f : X stretch(->>, size: #150%)_"surjective" Y $
+/// $ x stretch(harpoons.ltrb, size: #3em) y
+///     stretch(\[, size: #150%) z $
+/// ```
+#[elem(LayoutMath)]
+pub struct StretchElem {
+    /// The glyph to stretch.
+    #[required]
+    pub body: Content,
+
+    /// The size to stretch to, relative to the glyph's current size.
+    pub size: Smart<Rel<Length>>,
+
+    /// Whether the size should be relative to the width of any top or bottom
+    /// attachments.
+    ///
+    /// ```example
+    /// #set math.stretch(attach: false)
+    /// $ stretch(->, size: #250%)_"very long text" $
+    /// ```
+    #[default(true)]
+    pub attach: bool,
+}
+
+impl LayoutMath for Packed<StretchElem> {
+    #[typst_macros::time(name = "math.stretch", span = self.span())]
+    fn layout_math(&self, ctx: &mut MathContext, styles: StyleChain) -> SourceResult<()> {
+        let mut fragment = ctx.layout_into_fragment(self.body(), styles)?;
+        stretch_fragment(
+            ctx,
+            styles,
+            &mut fragment,
+            None,
+            None,
+            self.size(styles),
+            Abs::zero(),
+        );
+        ctx.push(fragment);
+        Ok(())
+    }
+}
+
+/// Attempts to stretch the given fragment by/to the amount given in stretch.
+pub(super) fn stretch_fragment(
+    ctx: &mut MathContext,
+    styles: StyleChain,
+    fragment: &mut MathFragment,
+    axis: Option<Axis>,
+    relative_to: Option<Abs>,
+    stretch: Smart<Rel<Length>>,
+    short_fall: Abs,
+) {
+    let glyph = match fragment {
+        MathFragment::Glyph(glyph) => glyph.clone(),
+        MathFragment::Variant(variant) => {
+            GlyphFragment::new(ctx, styles, variant.c, variant.span)
+        }
+        _ => return,
+    };
+
+    let axis = if let Some(stretch_axis) = axis {
+        stretch_axis
+    } else if let Some(stretch_axis) = is_stretchable(ctx, &glyph) {
+        stretch_axis
```

Comment:
looks like a candidate for `or_else`

Reply:
Surprised I missed that. Could've sworn I stared for a while at this and the docs to make it nicer aha.